packages/javascript/bh-shared-ui/src/components/FileUploadDialog/FileUploadDialog.tsx

```diff
@@ -15,68 +15,200 @@
+
+    useEffect(() => {
+        const filesHaveErrors = filesForIngest.filter((file) => file.errors).length > 0;
+        const filesAreUploading = filesForIngest.filter((file) => file.status === FileStatus.UPLOADING).length > 0;
+
+        if (filesHaveErrors || !filesForIngest.length) {
+            setSubmitDialogDisabled(true);
+        } else {
+            setSubmitDialogDisabled(false);
+        }
+
+        if (filesAreUploading) {
+            setUploadDialogDisabled(true);
+        } else {
+            setUploadDialogDisabled(false);
+        }
+    }, [filesForIngest]);
+
+    const handleRemoveFile = (index: number) => {
+        setFilesForIngest((prevFiles) => prevFiles.filter((_file, i) => i !== index));
+    };
+
+    const handleAppendFiles = (files: FileForIngest[]) => {
+        setFilesForIngest((prevFiles) => [...prevFiles, ...files]);
+    };
+
+    const updateStatusOfReadyFiles = (status: FileStatus) => {
+        setFilesForIngest((prevFiles) =>
+            prevFiles.map((file) => {
+                return {
+                    ...file,
+                    status: file.status === FileStatus.READY ? status : file.status,
+                };
+            })
+        );
+    };
+
+    const setNewFileStatus = (name: string, status: FileStatus) => {
+        setFilesForIngest((prevFiles) =>
+            prevFiles.map((file) => {
+                if (file.file.name === name) {
+                    return { ...file, status };
+                }
+                return file;
+            })
+        );
+    };
+
+    const setUploadFailureError = (name: string, error: string) => {
+        setNewFileStatus(name, FileStatus.FAILURE);
+
+        setFilesForIngest((prevFiles) =>
+            prevFiles.map((file) => {
+                if (file.file.name === name) {
+                    return { ...file, errors: [error] };
+                }
+                return file;
+            })
+        );
+    };
+
+    const handleUploadAllFiles = async () => {
+        updateStatusOfReadyFiles(FileStatus.UPLOADING);
+
+        try {
+            const response = await startUpload();
+            const currentIngestJobId = response?.data?.id?.toString();
+
+            // Counting errors manually here to avoid race conditions with react state updates
+            let errorCount = 0;
+
+            if (currentIngestJobId) {
+                for (const ingestFile of filesForIngest) {
+                    // Separate error handling so we can continue on when a file fails
+                    try {
+                        await uploadFile(currentIngestJobId, ingestFile);
+                    } catch (error) {
+                        errorCount += 1;
+                    }
+                    setNewFileStatus(ingestFile.file.name, FileStatus.DONE);
+                }
+            }
+            await finishUpload(currentIngestJobId);
+
+            logFinishedIngestJob(errorCount);
+        } catch (error) {
```

Comment:
I have a couple of questions around error handling:

1. Just a little clarification for me: whats going on with the start/finishUpload requests? Is it there a transaction type of functionality happening on the API where if one file errors, then no file gets uploaded? Or is that have to do with the file streaming?
2. if this function does error (specifically with start/finishUpload), should we also roll the `fileUploadStep` back to `CONFIRMATION` or `ADD_FILES` in the catch block? Each individual file upload has its own error handler, so Im thinking if the start/finishUpload request errors then nothing gets uploaded?

Reply:
Clarification: we send one request to start a job, one request each for each file we want to add to the job, and one request to close the job. If the job doesn't get closed, the job will eventually get marked as stale and discarded, but if a job gets closed and one or more of the files failed to ingest, my understanding is that the job will still run as expected.

I think it would be interesting to talk about what we want to do if the start or finish jobs fail. My thought would be instead of rolling them back a step, display an error and let them immediately retry, since a failure on the start or end requests is almost certainly an issue on our end. But there may be a case for bringing them back further in the process!

Reply:
Thank yee very much for the clarification! This definitely makes me think the current error handling should totally work.

What you said about start/finish failing makes sense. I think we should ship this as is and if we run into issues with start/finish then we can re-address the error flow with more real world use-cases.

Thanks again for the detailed response :)